local_adaptation/localadaptation_nonWF.slim

```diff
@@ -0,0 +1,153 @@
+    initializeGenomicElement(g1, asInteger(4e5), 6e5 - 1); 
+    initializeGenomicElement(g0, asInteger(6e5), 1e6 - 1);
+    initializeRecombinationRate(R);
+    // competition
+    initializeInteractionType(1, "xy", reciprocal=T, maxDistance=SC * 3);
+    i1.setInteractionFunction("n", 1.0, SC);
+    // mate choice
+    initializeInteractionType(2, "xy", reciprocal=T, maxDistance=SM * 3);
+    i2.setInteractionFunction("n", 1.0, SM);
+}
+
+mutationEffect(m2) { return 1.0; }
+1 first() {
+    sim.addSubpop("p1", asInteger(K * WIDTH * WIDTH));
+    p1.setSpatialBounds(c(0.0, 0.0, WIDTH, WIDTH));
+    p1.individuals.setSpatialPosition(p1.pointUniform(p1.individualCount));
+    p1.individuals.z = 0.0;
+    mapImage = Image(env_image);
+    
+    // mapImage.floatK for greyscale images, mapImage.floatG for RGB -- better to use gs
+    map = p1.defineSpatialMap("map1", "xy", 1.0 - mapImage.floatK,
+		  valueRange=c(0.0, 1.0), colors=c("red", "yellow"))
+		  .rescale(-1, 1);
+    defineConstant("OPTIMUM", map);
+}
+
+early() {
+    // Disperse offspring
+    offspring = p1.subsetIndividuals(maxAge=0);
+    p1.deviatePositions(offspring, "reprising", INF, "n", SD);
+    
+    // Measure local density and use it for density regulation
+    i1.evaluate(p1);
+    inds = p1.individuals;
+    competition = i1.localPopulationDensity(inds);
+    inds.fitnessScaling = 1 / (1 + RHO * competition);
+
+    // construct phenotypes and fitness effects from QTLs
+    inds = sim.subpopulations.individuals;
+    phenotype = inds.sumOfMutationsOfType(m2);
+    location = inds.spatialPosition;
+    optimum = OPTIMUM.mapValue(location);
+    
+    qtl_fit = 1 + dnorm(phenotype, optimum, SK);
+    qtl_fit = qtl_fit/max(qtl_fit);
+    
+    inds.fitnessScaling =   inds.fitnessScaling * qtl_fit;
+    inds.tagF = optimum;
+    
+    // color individuals according to phenotype
+    inds.color = OPTIMUM.mapColor(phenotype);
+        
+    if (p1.individualCount == 0) {
+        catn("Population went extinct! Ending the simulation.");
+        sim.simulationFinished();
+    }
+}
+
+2: first() {
+    // evaluate mate choice in preparation for reproduction
+    i2.evaluate(p1);
+    print(sim.chromosome.mutationRates);
+}
+
+first() {
+    // preparation for the reproduction() callback
+    i2.evaluate(p1);
+}
+
+reproduction() {
+    // spatial mate choice
+    mate = i2.drawByStrength(individual, 1);
+    if (mate.size()){;
+        subpop.addCrossed(individual, mate, count=rpois(1, FECUN));
+    }
+    return;
+}
+
+RUNTIME late() {
+    
+    // assign tag values to be preserved
+    inds = sortBy(sim.subpopulations.individuals, "pedigreeID");
+    phenotype = inds.sumOfMutationsOfType(m2);
+    location = inds.spatialPosition;
+    optimum = OPTIMUM.mapValue(location);
+    // record tag values and pedigree IDs in metadata
+    metadataDict = Dictionary("optimum", optimum, "phenotype", phenotype, "ids", inds.pedigreeID, "defaults", defaults);
```

Comment:
little bug here; "defaults", defaults needs to be removed for script to work, still thinking through how to add this to the Dict here without having to redefine the whole defaults dict....

Reply:
Sorry it took me a while to get back to this. The last couple weeks have been crazy! To answer this, yes! It was as simple as defining defaults in the initialize step to fix the issue. I guess since I had defined the defaults dict at the beginning I didn't think to define it as a constant, but that makes a lot of sense and seems to save correctly now!